### PR TITLE
Issue/bsp 468 fix security issue indom4j

### PR DESCRIPTION
**Pull request by billy-mccarthy**
Uses Maven 3 and Java 8.
Updates tests to work with new validation

### CODE REVIEW

File: src/main/java/org/dom4j/rule/Rule.java [hunk 1]
```diff
@@ -99,17 +99,17 @@ public int compareTo(Object that) {
      * @return DOCUMENT ME!
      */
     public int compareTo(Rule that) {
-        int answer = this.importPrecedence - that.importPrecedence;
+        double answer = this.importPrecedence - that.importPrecedence;
 
         if (answer == 0) {
-            answer = (int) Math.round(this.priority - that.priority);
+            answer = this.priority - that.priority;
```
Comment (billy-mccarthy) [inline src/main/java/org/dom4j/rule/Rule.java:105]: This was breaking an ordering test case as the rounding was rounding -0.5 to 0 meaning objects were being treated equally

File: src/test/java/org/dom4j/IteratorTest.java [hunk 2]
```diff
@@ -45,16 +45,16 @@ public void testElementCount() throws Exception {
...
         int elementSize = elements.size();
         assertTrue("Root has " + elementSize + " children", (elements != null)
                 && (elementSize == NUMELE));
     }
 
     public void testPlainIteration() throws Exception {
         Element root = iterDocument.getRootElement();
-        List elements = root.elements("iterator test");
-        Iterator iter = root.elementIterator("iterator test");
+        List elements = root.elements("iteratortest");
```
Comment (billy-mccarthy) [inline src/test/java/org/dom4j/IteratorTest.java:56]: spaces are now not valid under changes made to QName

File: src/test/java/org/dom4j/io/StaxTest.java [hunk 3]
```diff
@@ -50,7 +51,7 @@ public void testEncoding() {
         try {
             File file = getFile("/xml/russArticle.xml");
             STAXEventReader xmlReader = new STAXEventReader();
-            Document doc = xmlReader.readDocument(new FileReader(file));
+            Document doc = xmlReader.readDocument(new FileInputStream(file));
```
Comment (billy-mccarthy) [inline src/test/java/org/dom4j/io/StaxTest.java:54]: Using a reader means the underlying encoding is not read has to use input stream as per https://stackoverflow.com/questions/3482494/howto-let-the-sax-parser-determine-the-encoding-from-the-xml-declaration

File: src/test/java/org/dom4j/io/StaxTest.java [hunk 4]
```diff
@@ -61,7 +62,7 @@ public void testEncoding() {
 
             String output = writer.toString();
             String xmlDecl = output.substring(0, output.indexOf("?>") + 2);
-            String expected = "<?xml version=\'1.0\' encoding=\'koi8-r\'?>";
+            String expected = "<?xml version=\"1.0\" encoding=\"koi8-r\"?>";
```
Comment (billy-mccarthy) [inline src/test/java/org/dom4j/io/StaxTest.java:65]: This now matches the xml

File: xml/test/schema/test.xsd [hunk 5]
```diff
@@ -1 +1,45 @@
.../>        <element name="gDayTag" type="gDay"/>    	<element name="gMonthTag" type="gMonth"/>    	<element name="booleanTag" type="boolean"/>    	<element name="base64BinaryTag" type="base64Binary"/>    	<element name="hexBinaryTag" type="hexBinary"/>	    	<element name="floatTag" type="float"/>    	<element name="doubleTag" type="double"/>	    	<element name="decimalTag" type="decimal"/>    	<element name="integerTag" type="integer"/>	    	<element name="nonPositiveIntegerTag" type="nonPositiveInteger"/>    	<element name="negativeIntegerTag" type="negativeInteger"/>	    	<element name="longTag" type="long"/>    	<element name="intTag" type="int"/>    	<element name="shortTag" type="short"/>    	<element name="byteTag" type="byte"/>	    	<element name="nonNegativeIntegerTag" type="nonNegativeInteger"/>    	<element name="unsignedLongTag" type="unsignedLong"/>    	<element name="unsignedIntTag" type="unsignedInt"/>    	<element name="unsignedShortTag" type="unsignedShort"/>    	<element name="unsignedByteTag" type="unsignedByte"/>    	<element name="positiveIntegerTag" type="positiveInteger"/>	</sequence>    <attribute name="id" type="string" />  </complexType>	 </element></schema>
```
Comment (billy-mccarthy) [inline xml/test/schema/test.xsd:1]: Reformatted to improve readability of xsd

File: src/test/java/org/dom4j/XPathTest.java [hunk 6]
```diff
@@ -22,7 +22,6 @@
  */
 public class XPathTest extends AbstractTestCase {
     protected static String[] paths = {
-            ".",
```
Comment (billy-mccarthy) [inline src/test/java/org/dom4j/XPathTest.java:25]: This is an invalid character to be read by QName

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+*.iml
+.idea
+.clover
+.classpath
+.project
+.settings
+*.iml
+*.ipr
+*.iws
+.idea
```
Comment (andrewswan) [inline .gitignore:19]: Normally each developer would put these IDE-specific exclusions into their own global `.gitignore` file, to apply across their whole machine.

File: README.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# dom4j
+dom4j is an open source framework for processing XML which is integrated with XPath and fully supports DOM, SAX, JAXP and the Java platform such as Java 2 Collections.
+
+Version 1.6.2 Bug fix release.
+
+This is an Atlassian fork of dom4j 
+
+See original source here https://github.com/dom4j/dom4j
+
+Track changes on on the Atlassian fork here https://github.com/atlassian/dom4j
```
Comment (andrewswan) [inline README.md]: "on on"

File: pom.xml [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>3.8.2</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+    <build>
+        <sourceDirectory>src/main/java</sourceDirectory>
+        <testSourceDirectory>src/test/java</testSourceDirectory>
```
Comment (andrewswan) [inline pom.xml]: Are these two values not the default?

File: pom.xml [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        <plugins>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <debug>true</debug>
+                    <optimize>false</optimize>
+                    <showDeprecation>true</showDeprecation>
+                    <source>1.6</source>
+                    <target>1.6</target>
```
Comment (andrewswan) [inline pom.xml]: Not 1.8?

File: src/main/java/org/dom4j/Element.java [hunk 11]
```diff
@@ -643,7 +643,7 @@
      *            is the attribute's value
      * 
      * @deprecated As of version 0.5. Please use {@link
-     *             #addAttribute(QName,String)} instead. WILL BE REMOVED IN
+     *             #addAttribute(Name,String)} instead. WILL BE REMOVED IN
      *             dom4j-1.6 !!
```
Comment (andrewswan) [inline src/main/java/org/dom4j/Element.java:647]: Oh will it now?

File: src/main/java/org/dom4j/dom/DOMDocument.java [hunk 12]
```diff
@@ -15,12 +15,16 @@
...
+import org.w3c.dom.DOMConfiguration;
 import org.w3c.dom.DOMException;
 import org.w3c.dom.Document;
 import org.w3c.dom.EntityReference;
 import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
 import org.w3c.dom.NodeList;
 import org.w3c.dom.ProcessingInstruction;
+import org.w3c.dom.UserDataHandler;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
Comment (andrewswan) [inline src/main/java/org/dom4j/dom/DOMDocument.java]: Isn't this class specific to the Oracle JDK? Did you use it on purpose?

File: src/main/java/org/dom4j/dom/DOMElement.java [hunk 13]
```diff
@@ -16,11 +16,15 @@
...
 
+import org.w3c.dom.Attr;
 import org.w3c.dom.DOMException;
 import org.w3c.dom.Document;
 import org.w3c.dom.NamedNodeMap;
 import org.w3c.dom.Node;
 import org.w3c.dom.NodeList;
+import org.w3c.dom.TypeInfo;
+import org.w3c.dom.UserDataHandler;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
Comment (andrewswan) [inline src/main/java/org/dom4j/dom/DOMElement.java]: Same here.

File: src/main/java/org/dom4j/dom/DOMEntityReference.java [hunk 14]
```diff
@@ -13,7 +13,10 @@
 import org.w3c.dom.DOMException;
 import org.w3c.dom.Document;
 import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
 import org.w3c.dom.NodeList;
+import org.w3c.dom.UserDataHandler;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
Comment (andrewswan) [inline src/main/java/org/dom4j/dom/DOMEntityReference.java]: And here.

File: src/main/java/org/dom4j/rule/Rule.java [hunk 15]
```diff
@@ -99,17 +99,17 @@ public int compareTo(Object that) {
...
 
             if (answer == 0) {
                 answer = this.appearenceCount - that.appearenceCount;
             }
         }
 
-        return answer;
+        return answer < 0 ? -1 : answer > 0 ? 1 : 0;
```
Comment (andrewswan) [inline src/main/java/org/dom4j/rule/Rule.java:112]: Why this change? The original one satisfies the contract of `compareTo`.

File: src/test/java/org/dom4j/AbstractTestCase.java [hunk 16]
```diff
@@ -23,14 +23,14 @@
  * @author <a href="mailto:jstrachan@apache.org">James Strachan </a>
  * @version $Revision$
  */
-public class AbstractTestCase extends TestCase {
+public abstract class AbstractTestCase extends TestCase {
     protected Document document;
 
-    protected AbstractTestCase() {
+    public AbstractTestCase() {
```
Comment (andrewswan) [inline src/test/java/org/dom4j/AbstractTestCase.java:29]: Why make the constructors `public`? They can only be called by subclasses anyway, so they should be `protected`.

File: src/test/java/org/dom4j/AllowedCharsTest.java [hunk 17]
```diff
@@ -16,83 +16,90 @@ public void testLocalName() {
...
     }
 
     public void testQname() {
-        try {
-            QName.get("element", "http://example.com/namespace");
-            QName.get("ns:element", "http://example.com/namespace");
-            fail("Expected exception IllegalArgumentException.class but none occured")
-        } catch (IllegalArgumentException.class){}
+        QName.get("element", "http://example.com/namespace");
+        QName.get("ns:element", "http://example.com/namespace");
```
Comment (andrewswan) [inline src/test/java/org/dom4j/AllowedCharsTest.java:26]: So these are now legal arguments, whereas before they were not?

File: pom.xml [hunk 18]
```diff
@@ -0,0 +1,146 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd"
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>dom4j</groupId>
+    <artifactId>dom4j</artifactId>
+    <version>1.6.1</version>
```
Comment (billy-mccarthy) [inline pom.xml]: need to update to 1.6.2-SNAPSHOT

File: pom.xml (hunk 10 above)
Comment (billy-mccarthy) [inline pom.xml]: Bump to 1.8

File: src/test/java/org/dom4j/datatype/SchemaParseTest.java [hunk 19]
```diff
@@ -14,7 +14,7 @@
 import org.dom4j.io.SAXReader;
 
 /**
- * Tests the SchemaParser based on a test case provided by Luis Pe�a S�nchez
+ * Tests the SchemaParser based on a test case provided by Luis Peta Sanchez
```
Comment (andrewswan) [inline src/test/java/org/dom4j/datatype/SchemaParseTest.java]: ?

**Review by andrewswan — CHANGES_REQUESTED** [review]

File: src/main/java/org/dom4j/rule/Rule.java (hunk 1 above)
Comment (ablekhman) [reply src/main/java/org/dom4j/rule/Rule.java:105]: You may never be able to satisfy this condition, since comparing floating point numbers to exact values rarely produces expected results. You need to compare the difference between them to some predefined tolerance value.

Comment (billy-mccarthy) [reply src/main/java/org/dom4j/rule/Rule.java:105]: The test case that was breaking now works as a result of this change so I believe it does the job

File: pom.xml (hunk 9 above)
Comment (billy-mccarthy) [reply pom.xml]: This pom was generated when I ran mvn one:convert happy to remove though

File: src/main/java/org/dom4j/dom/DOMDocument.java (hunk 12 above)
Comment (billy-mccarthy) [reply src/main/java/org/dom4j/dom/DOMDocument.java]: I wasn't sure what exception to throw, you may be right any suggestions? RuntimeException?

File: src/main/java/org/dom4j/rule/Rule.java (hunk 15 above)
Comment (billy-mccarthy) [reply src/main/java/org/dom4j/rule/Rule.java:112]: The method needs to return an int and upon casting -0.5 to an int the precision was lost and therefore the objects were incorrectly being sorted. This was required to pass a test

File: src/test/java/org/dom4j/AbstractTestCase.java (hunk 16 above)
Comment (billy-mccarthy) [reply src/test/java/org/dom4j/AbstractTestCase.java:29]: They were showing as failed tests when running all tests in the package due to a warning from junit I believe

File: src/test/java/org/dom4j/AllowedCharsTest.java (hunk 17 above)
Comment (billy-mccarthy) [reply src/test/java/org/dom4j/AllowedCharsTest.java:26]: This was a mistake in the original PR https://github.com/atlassian/dom4j/pull/1 I incorrectly converted this case

File: src/test/java/org/dom4j/datatype/SchemaParseTest.java (hunk 19 above)
Comment (billy-mccarthy) [reply src/test/java/org/dom4j/datatype/SchemaParseTest.java]: I'll remove this comment altogether as adds no value
